### PR TITLE
Implement executionhook controller

### PR DESCRIPTION
**Pull request by ashish-amarnath**
Implement executionhook controller

Signed-off-by: Ashish Amarnath ashisham@vmware.com

### CODE REVIEW

File: api/v1alpha1/executionhook_types.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+	// Succeed is set to true when the action is executed in the container successfully.
+	// It will be set to false if the action cannot be executed successfully after
+	// ActionTimeoutSeconds passes.
+	// +optional
+	Succeed *bool `json:"actionSucceed,omitempty" protobuf:"varint,4,opt,name=actionSucceed"`
+
+	// The last error encountered when executing the action. The hook controller might
+	// update this field each time it retries the execution.
+	// +optional
+	Error *HookError `json:"error,omitempty" protobuf:"bytes,5,opt,name=error"`
```
Comment (kow3ns) [inline api/v1alpha1/executionhook_types.go:122]: There can be only one error? That is not an unreasonable assumption. However, this pattern seems to be similar to the Conditions pattern implemented by most built in types. Is there a reason to go with a singular error over Status.Conditions?

File: api/v1alpha1/executionhook_types.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+	Timestamp *metav1.Time `json:"timestamp,omitempty" protobuf:"bytes,4,opt,name=timestamp"`
+}
+
+// ErrorType defines the type of error occurred from hook execution.
+type ErrorType string
+
+// More error types could be added, e.g., Forbidden, Unauthorized, AlreadyInProgress, etc.
+const (
+	// The execution hook times out
+	Timeout ErrorType = "Timeout"
```
Comment (kow3ns) [inline api/v1alpha1/executionhook_types.go:150]: Another reason to consider using the Conditions pattern. This proposal might encourage end users to build a state machine off the error conditions (which as you point out in the comment) are likely to be extended and change. Across the rest of the API we generally avoid this as we've learned that its brittle.

**Review by kow3ns — APPROVED** [review]
Looks good overall. One comment on the API and the testing is a bit sparse, but its a great start!

File: api/v1alpha1/executionhook_types.go (hunk 1 above)
Comment (ashish-amarnath) [reply api/v1alpha1/executionhook_types.go:122]: The expectation is that the error represents the error condition encountered executing the hook action on a specific container in a pod. So there is a 1:1 mapping between where the hook action was executed (Pod,Container) and an error condition, if any. For that reason, using a singular error was thought of as a better fit. However, using `Status.Conditions` would be more appropriate if/ when we aggregate the errors into the status of the `ExecutionHook` object.

File: api/v1alpha1/executionhook_types.go (hunk 2 above)
Comment (ashish-amarnath) [reply api/v1alpha1/executionhook_types.go:150]: Thanks for the suggestion. I will revisit this as I iterate on the implementation. WDYT?

File: api/v1alpha1/executionhook_types.go (hunk 1 above)
Comment (xing-yang) [reply api/v1alpha1/executionhook_types.go:122]: I think we were given recommendations not to use Conditions during the API review, although I couldn't remember the reason behind that.   @thockin @jingxu97 do you want to comment on this?

File: api/v1alpha1/executionhook_types.go (hunk 2 above)
Comment (xing-yang) [reply api/v1alpha1/executionhook_types.go:150]: @kow3ns If Conditions pattern is used, end user can still add more condition types?

**Comment by xing-yang** [comment]
Hi @ashish-amarnath, should this PR be dependent on https://github.com/kubernetes-sigs/execution-hook/pull/5?

**Comment by ashish-amarnath** [comment]
@xing-yang PR #4 (this PR)  is the super set of changes in PR #5.
PR #5 was created to breakup the API changes from the controller implementation and unit tests, as it was suggested during last weeks meeting.

I would prefer getting PR #4 (this PR)  merged and closing PR #5

**Pull request closed without merging**